Add TimmWrapper

qubvel opened the pull request:
# What does this PR do?

Adds a TimmWrapper set of classes such that timm models can be loaded in as transformer models into the library. 

Continue of 
 - https://github.com/huggingface/transformers/pull/33687

### General Usage

```py
import torch
from urllib.request import urlopen
from PIL import Image
from transformers import AutoConfig, AutoModelForImageClassification, AutoImageProcessor

checkpoint = "timm/resnet50.a1_in1k"
img = Image.open(urlopen(
    'https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/beignets-task-guide.png'
))

image_processor = AutoImageProcessor.from_pretrained(checkpoint)
inputs = image_processor(img, return_tensors="pt")
model = AutoModelForImageClassification.from_pretrained(checkpoint)

with torch.no_grad():
    logits = model(**inputs).logits

top5_probabilities, top5_class_indices = torch.topk(logits.softmax(dim=1) * 100, k=5)
```


### Pipeline

Timm models can now be used in the image classification (if a classification model) and image feature extraction pipelines

```py
import torch
from urllib.request import urlopen
from PIL import Image

from transformers import pipeline

img = Image.open(urlopen(
    'https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/beignets-task-guide.png'
))
pipe = pipeline("image-classification", model="timm/resnet18.a1_in1k")
print(pipe(img))
```

### Trainer

Timm models can now be loaded and trained with the trainer class. 

Example model trained with the trainer running the script command below:
https://huggingface.co/qubvel-hf/vit-base-beans

```
python run_image_classification.py \                
    --dataset_name beans \
    --output_dir ./beans_outputs/ \
    --remove_unused_columns False \
    --label_column_name labels \
    --do_train \
    --do_eval \
    --push_to_hub \
    --push_to_hub_model_id vit-base-beans \
    --learning_rate 2e-5 \
    --num_train_epochs 5 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 8 \
    --logging_strategy steps \
    --logging_steps 10 \
    --eval_strategy epoch \
    --save_strategy epoch \
    --load_best_model_at_end True \
    --save_total_limit 3 \
    --seed 1337 \
    --model_name_or_path timm/resnet18.a1_in1k \
    --ignore_mismatched_sizes
```

### Other features enabled

 - [x] Device map: 
 ```python
model = TimmWrapperForImageClassification.from_pretrained(checkpoint, device_map="auto")
 ```
 - [x] Torch dtype:
 ```python
model = TimmWrapperForImageClassification.from_pretrained(checkpoint, torch_dtype="bfloat16")
 ```
 - [x] Quantization:
 ```python
model = TimmWrapperForImageClassification.from_pretrained(checkpoint, load_in_4bit=True)
 ```
 - [x] Intermediate hidden states: `output_hidden_states=True` or `output_hidden_states=[1, 2, 3]` (to select specific hidden states)
```python
model = TimmWrapperForImageClassification.from_pretrained(checkpoint)
output = model(**intpus, output_hidden_states=True)
 ```
 
  - [x] Transformers TImmWrapper checkpoints are compatible with timm:
```python
model = timm.create_model("hf-hub:qubvel-hf/vit-base-beans", pretrained=True)
 ```
 
### TODO

 - [x] Gamma/beta renaming issue
 - [x] Update timm in CI 0.9.6 -> 1.0.11 to enable `output_hidden_states` tests
   - CI for slow-run takes longer to update images
 - [x] Weights are loaded by `transformers` instead of `timm`, which architectures are affected?
   - mostly OpenCLIP weights are affected (60-70 checkpoints), see the comment https://github.com/huggingface/transformers/pull/34564#issuecomment-2491972650
 - [x] Tests for image processor

amyeroberts committed:
Add files

amyeroberts committed:
Init

amyeroberts committed:
Add TimmWrapperModel

amyeroberts committed:
Fix up

amyeroberts committed:
Some fixes

amyeroberts committed:
Fix up

amyeroberts committed:
Remove old file

amyeroberts committed:
Sort out import orders

amyeroberts committed:
Fix some model loading

amyeroberts committed:
Compatible with pipeline and trainer

amyeroberts committed:
Fix up

amyeroberts committed:
Delete test_timm_model_1/config.json

amyeroberts committed:
Remove accidentally commited files

amyeroberts committed:
Delete src/transformers/models/modeling_timm_wrapper.py

amyeroberts committed:
Remove empty imports; fix transformations applied

amyeroberts committed:
Tidy up

amyeroberts committed:
Add image classifcation model to special cases

amyeroberts committed:
Create pretrained model; enable device_map='auto'

amyeroberts committed:
Enable most tests; fix init order

amyeroberts committed:
Sort imports

amyeroberts committed:
[run-slow] timm_wrapper

qubvel committed:
Pass num_classes into timm.create_model

qubvel committed:
Remove train transforms from image processor

qubvel committed:
Update timm creation with pretrained=False

qubvel committed:
Fix gamma/beta issue for timm models

qubvel committed:
Fixing gamma and beta renaming for timm models

qubvel committed:
Simplify config and model creation

qubvel committed:
Remove attn_implementation diff

qubvel committed:
Fixup

qubvel committed:
Docstrings

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_34564). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel committed:
Fix warning msg text according to test case

qubvel committed:
Fix device_map auto

qubvel committed:
Set dtype and device for pixel_values in forward

qubvel committed:
Enable output hidden states

qubvel committed:
Enable tests for hidden_states and model parallel

qubvel committed:
Remove default scriptable arg

qubvel committed:
Refactor inner model

qubvel committed:
Update timm version

qubvel commented on src/transformers/modeling_utils.py:
There are 2 changes in this file:

1. state_dict keys renaming moved into separate method to be able to override it for TimmWrapper (disable gamma/beta renaming + add prefix)
2. metadata is None for timm checkpoints -> assuming these are pytorch checkpoints

qubvel commented on src/transformers/models/auto/image_processing_auto.py:
```diff
@@ -412,6 +414,10 @@ def from_pretrained(cls, pretrained_model_name_or_path, *inputs, **kwargs):
         use_fast = kwargs.pop("use_fast", None)
         trust_remote_code = kwargs.pop("trust_remote_code", None)
         kwargs["_from_auto"] = True
+        default_image_processor_filename = (
+            "config.json" if is_timm_checkpoint(pretrained_model_name_or_path) else IMAGE_PROCESSOR_NAME
+        )
+        kwargs["image_processor_filename"] = kwargs.get("image_processor_filename", default_image_processor_filename)
```
timm checkpoints store image processor config in `config.json`

qubvel commented:
@rwightman could you please make the first review in case you have bandwidth

rwightman commented:
@qubvel I'm starting to work through it now, wanted to get eval working to check some familiar models and just wasted way too much time realizing I needed `--remove_unused_columns False ` for anything to work at all :/ ... that's a really poor setup when most datasets have an 'image' column and not a 'pixel_values' column :/

Annnyways, first pass of the code things looked sane but need to spend some time looking closer at the details and testing some cases.

rwightman commented:
A few high level q...

* Does 'Wrapper' add any worthwhile value/info in the name vs
  * `TimmPreTrainedModel(PreTrainedModel)`
  * `TimmModel(TimmPretrainedModel)`
  * `TimmModelForImageClassification(..)`

* Is there a reason TimmWrapperModel has `.timm_model`, instead of something more generic like `.model`

* Any reservations in changing TimmWrapperModelForImageClassification to not use TimmWrapperModel? There are a few issues with the handling details for classifier, possible optimizations for forward call sequence and it'd probably be a bit cleaner to just duplicate a bit of redunant code and keep the two impl separate and a bit different.

* I thought we were going to set pretained=True so timm can load the weights, their are a number of weight adaptation / translation things that don't run if this isn't use, cannot change num_classes cleanly for instance. 

* What happens if we try to push these models to the hub? Do they get uploaded/written in a form that timm can read?

qubvel commented:
> Does 'Wrapper' add any worthwhile value/info in the name vs
TimmPreTrainedModel(PreTrainedModel)
TimmModel(TimmPretrainedModel)
TimmModelForImageClassification(..)

Left it as it was in the previous PR, however, `TimmModelForImageClassification` sounds better to me, I can rename it

> Is there a reason TimmWrapperModel has .timm_model, instead of something more generic like .model

The prefix "timm_model" is unique and is used in certain tests to identify when weights come from a timm model. It's also utilized in the `_fix_state_dict_key` method to determine whether to add the prefix when loading weights from the original checkpoint. For these reasons, I would prefer to keep it as "timm_model"

> Any reservations in changing TimmWrapperModelForImageClassification to not use TimmWrapperModel? There are a few issues with the handling details for classifier, possible optimizations for forward call sequence and it'd probably be a bit cleaner to just duplicate a bit of redunant code and keep the two impl separate and a bit different.

Originally, this was implemented without `TimmWrapperModel` in `TimmWrapperModelForImageClassification`, but I introduced it to reduce code repetition. This approach also aligns better with common patterns in the transformers repo. Could you provide more details on the issues with the classifier? In any case, the code will remain the same for both models if we aim to maintain `output_hidden_states` functionality across them.

> I thought we were going to set pretained=True so timm can load the weights, their are a number of weight adaptation / translation things that don't run if this isn't use, cannot change num_classes cleanly for instance.


I left a comment in the thread about this. We use transformers for weight loading to leverage features like device_map, torch_dtype, and quantization. I’m also unsure how to disable weight loading through transformers if it’s handled by timm, as I haven’t seen any examples of this in the repo. I can dig into it further if needed.

Do you have an estimate of how many models involve weight renaming? Is there a way to update checkpoints without breaking older versions of timm? Alternatively, could we manage similar weight renaming directly in transformers? (Though I think this approach may be less robust.)

> What happens if we try to push these models to the hub? Do they get uploaded/written in a form that timm can read?
I’m not sure if it's currently compatible, but it would be great to enable it! The config should be compatible, however, the weights state dict will have the "model.timm_model." prefix. I can look into removing this prefix before saving. I will try to enable it and add a separate test for a few checkpoints. Thanks for bringing this up!

rwightman commented:
> Originally, this was implemented without `TimmWrapperModel` in `TimmWrapperModelForImageClassification`, but I introduced it to reduce code repetition. This approach also aligns better with common patterns in the transformers repo. Could you provide more details on the issues with the classifier? In any case, the code will remain the same for both models if we aim to maintain `output_hidden_states` functionality across them.

I feel in this case there is a difference in alignment with other models, because both ImageClassification and base model wrap timm model instances that differ, instead of adding their own head to the same timm model. I see some options, depending on the mix of hidden state flags, head vs no head where it'd be appropriate to go through different forward calls, where more than just one argument might be appropriate to change on creation, etc.

Also, thinking about the future and other tasks. I feel there is a high probability say for supporting native timm object detection more flexibility is desired so it'd be safer to have it uncoupled.

That and there's a resisitance to significant changes in transformers after something is in there, so feel it's safer to leave uncoupled to have additional flexibility and avoid being stuck with tricky decisions that might impact timm moving forward.


And last motivation, there's a

rwightman commented:
> > I thought we were going to set pretained=True so timm can load the weights, their are a number of weight adaptation / translation things that don't run if this isn't use, cannot change num_classes cleanly for instance.
> 
> I left a comment in the thread about this. We use transformers for weight loading to leverage features like device_map, torch_dtype, and quantization. I’m also unsure how to disable weight loading through transformers if it’s handled by timm, as I haven’t seen any examples of this in the repo. I can dig into it further if needed.
> 
> Do you have an estimate of how many models involve weight renaming? Is there a way to update checkpoints without breaking older versions of timm? Alternatively, could we manage similar weight renaming directly in transformers? (Though I think this approach may be less robust.)

Aside from renaming, it just doesn't work right now, you can't load weights for a model if the head size changes for a new classification task. The wrapper only works if you use the imagenet classifier.

I realize timm doesn't have the dtype, lazy init features but it's better to have it work I feel. Can potentially look at supporting some of that through timm. It hasn't been a priority as there aren't too many very large models in timm.

If there's no way to do use pretrained=True on creation then will probably need to figure out how to add & call a method in timm once transformers has the state_dict and before it's loaded into the model, not sure if there's a spot for that?

rwightman commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,215 @@
+        Empty init weights function to ensure compatibility of the class in the library.
+        """
+        if isinstance(module, (nn.Linear)):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
```
it's not clear why this is here, what it's attempting to do. timm has model specific init fns though they aren't separately callable right now, doing something like this that could overwrite timm defaults would change model behaviour

qubvel replied:
Added this to initialize classifier, without this weights are not properly initialized, probably due to how the model is created in transformers

```python
from transformers import TimmWrapperForImageClassification

# --------------
# With init
# --------------

model = TimmWrapperForImageClassification.from_pretrained("timm/resnet18.a1_in1k", num_labels=10, ignore_mismatched_sizes=True)
print(model.timm_model.fc.weight[:3, :3])
# tensor([[-0.2117, -0.2422, -0.2540],
#         [-0.1106, -0.1856, -0.0152],
#         [-0.3430, -0.6446, -0.0530]], grad_fn=<SliceBackward0>)


# --------------
# Without init
# --------------

# patch with empty init weight function to check
def empty_init(self, module):
    pass
TimmWrapperForImageClassification._init_weights = empty_init

model = TimmWrapperForImageClassification.from_pretrained("timm/resnet18.a1_in1k", num_labels=10, ignore_mismatched_sizes=True)
print(model.timm_model.fc.weight[:3, :3])
# tensor([[0., 0., 0.],
#         [0., 0., 0.],
#         [0., 0., 0.]], grad_fn=<SliceBackward0>)
```

Ideally, we should get rid of this, but it's not common for transformers to load external models, so it might require more code changes. For now, its a simple fix to enable model loading with initialized classifier if shapes are mismatched.

rwightman commented on src/transformers/utils/generic.py:
```diff
@@ -867,3 +869,51 @@ class LossKwargs(TypedDict, total=False):
+
+    if is_dir and os.path.exists(os.path.join(pretrained_model_name_or_path, IMAGE_PROCESSOR_NAME)):
+        # timm models don't have a preprocessor_config.json file saved out
+        return False
```
don't think absence of a file is a good check

qubvel committed:
Fix _find_mismatched_keys function

qubvel committed:
Change inheritance for Classification model (fix weights loading with device_map)

qubvel commented:
> Aside from renaming, it just doesn't work right now, you can't load weights for a model if the head size changes for a new classification task. The wrapper only works if you use the imagenet classifier.

Fixed!

qubvel commented:
> If there's no way to do use pretrained=True on creation then will probably need to figure out how to add & call a method in timm once transformers has the state_dict and before it's loaded into the model, not sure if there's a spot for such a call in transformers?

If there would be any timm model/class-specific function that fixes state_dict I suppose we can call it before loading weights, similar to `_fix_state_dict_key` in the current implementation. However, it will be only supported for newer timm versions

rwightman commented:
Thinking about this a bit more, there is another issue with the hub / transformers first weight loading. `timm` wasn't originally hub first, so the library itself is still the primary source of truth for some models, doing a hub based load won't work.

Example this model https://huggingface.co/laion/CLIP-ViT-B-16-laion2B-s34B-b88K ... is an OpenCLIP first model, but timm can load it if you use the model name 'vit_base_patch16_clip_224.laion2b' 

Indeed I feel that the wrapper should support all timm model names that work in timm, but right now if the model isn't on the hub w/ a timm primary config it isn't useable. Ideally it should work with both a hub model name OR any timm model name. The timm model name would require timm do the pretrained loading to resolve any translation to other hub name or weight source.

Some very popular examples of this

https://github.com/huggingface/pytorch-image-models/blob/51ac8d2efb926c6b7c34eeb1dc52bcf57999e2de/timm/models/vision_transformer.py#L1580-L1716

rwightman commented:
> > Aside from renaming, it just doesn't work right now, you can't load weights for a model if the head size changes for a new classification task. The wrapper only works if you use the imagenet classifier.
> 
> Fixed!

Yay, I was able to run fine-tune using run_image_classification.py after this fix. An observation the output files for that script aren't directly useable to push to hub.

There is no config matching timm format or name, there is a config output by the image preprocessor save process to a different filename that's a jumble of the timm config. 

Also, the state dict for the model has the timm_model prefix so it's not loadable in timm. Is there anyway to remove that prefix in the checkpoints? this would also make it more seamless doing local dir loads if someone had timm weights already and config files checked out.

qubvel committed:
Minor bugfix

qubvel committed:
Disable save pretrained for image processor

qubvel committed:
Rename hook method for loaded keys correction

qubvel committed:
Rename state dict keys on save, remove `timm_model` prefix, make checkpoint compatible with `timm`

qubvel committed:
Managing num_labels <-> num_classes attributes

qubvel committed:
Enable loading checkpoints in Trainer to resume training

qubvel commented:
Hi @rwightman, 

I've addressed some comments and pushed the following fixes:  

Config and weights are now compatible with native timm models.  
  - Removed the `timm_model` prefix.  
  - Disabled processor config saving, the configuration is now kept in `config.json`, consistent with the original timm checkpoint format.  

Here’s an example of a fine-tuned **TimmWrapper model** using `run_classification` that can be loaded directly into timm:  

```python
model = timm.create_model("hf-hub:qubvel-hf/vit-base-beans", pretrained=True)
```  

Regarding OpenCLIP models:  
Would it be possible to provide compatible copies of these checkpoints under the `timm/` org on the Hub? Specifically, reconverting them to the appropriate format.  

Thanks!

qubvel commented:
I've also noticed that some models from the provided [list](https://github.com/huggingface/pytorch-image-models/blob/51ac8d2efb926c6b7c34eeb1dc52bcf57999e2de/timm/models/vision_transformer.py#L1580-L1716) have transformers-like `config.json` along with `open_clip_config.json`, which allows them to be loaded with `CLIPForImageClassification`. Would it be possible to update other checkpoints similarly? We may confuse the users a bit, but updating model cards should help

qubvel committed:
Update error message for output_hidden_states

qubvel committed:
Add output hidden states test

qubvel committed:
Decouple base and classification models

qubvel committed:
Add more test cases

qubvel committed:
Add save-load-to-timm test

qubvel committed:
Merge branch 'main' into timm-wrapper

qubvel committed:
Fix test name

qubvel committed:
Fixup

qubvel commented:
- Saved checkpoints are now timm-compatible
- Removed inheritance from the `ForImageClassification` model and added additional tests
- The renaming of `TimmWrapperModel` to `TimmModel` is not yet done. I'm uncertain if this might cause issues, as the model folder would need to be named `timm`, which could potentially create confusion with the `timm` package

@rwightman, could you please review when you have the bandwidth? also please let me know what you think about uploading transformers compatible checkpoints to new or existing repos on the hub (see comments above)

rwightman commented:
@qubvel thanks, I'll dig in and check out the latest updates tomorrow.

rwightman commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,254 @@
+            last_hidden_state=last_hidden_state,
+            pooler_output=pooler_output,
+            hidden_states=hidden_states,
+        )
```
I'm not too familiar with the transformer norms, but what do you normally do in the case where you want a BaseModelOutput without pooling? Is there a normal argument you pass to enable / disable pooling? Or are we supposed to have a TimeWrapperWithPooling and TimmWrapperWithoutPooling and that's done by what class you create? People use both and imagine don't want to pool if it's not necessary for their use case...

qubvel replied:
is the last_hidden_state would be equivalent to no pooled output? 

where the last hidden state is:
```python
last_hidden_state, hidden_states = self.timm_model.forward_intermediates(pixel_values, **kwargs)
# or
last_hidden_state = self.timm_model.forward_features(pixel_values, **kwargs)
```

rwightman commented:
A user of the timm backbone was complaining about the transformers weight handling wrt to 'gamma', etc .. can the fixes for this use case be extended to the backbone easily?

rwightman commented on tests/models/timm_wrapper/test_modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,364 @@
+            "hrnet_w18.ms_aug_in1k",
+        ]
+        image = prepare_img()
+
```
Looks like these tests are focused on models with classification heads with a classification wrapper...

Other scenarios to consider
* timm model w/ classification head used with TimmWrapperModel
* timm model w/o classification head use with TimmWrapperModel
* timm model w/o classification head use with TimmWrapperForImageClassification  (head needs to be added by wrapper)

Some models wihout heads to check
* vit_base_patch16_siglip_224.webli ... this is OpenCLIP model w/o timm config in hub though https://huggingface.co/timm/ViT-B-16-SigLIP 
* vit_small_patch16_224.dino

qubvel replied:
The `TimmWrapperModel` represents a model without a classification head, while the `TimmWrapperForImageClassification` includes a classification head. The wrapper itself does not add any heads, instead, the classification head is provided entirely by the original `timm` model. Such a behavior indeed is different from other `transformers` models, however, it allows load weights to classification head with no issues + make weights timm-backward-compatible.

rwightman commented:
> * Saved checkpoints are now timm-compatible
> * Removed inheritance from the `ForImageClassification` model and added additional tests
> * The renaming of `TimmWrapperModel` to `TimmModel` is not yet done. I'm uncertain if this might cause issues, as the model folder would need to be named `timm`, which could potentially create confusion with the `timm` package, maybe it's safer to have it as timm_wrapper
> 
> @rwightman, could you please review when you have the bandwidth? also please let me know what you think about uploading transformers compatible checkpoints to new or existing repos on the hub (see comments above)

Good point re the transformer naming resulting in a folder named just 'timm', yeah not a good idea, let's stick with wrapper naming as is. Parallels timm_backbone then too.

qubvel commented:
> A user of the timm backbone was complaining about the transformers weight handling wrt to 'gamma', etc .. can the fixes for this use case be extended to the backbone easily?

yeah, probably the same trick will work, can have a look once this PR merged

rwightman commented:
@qubvel just noticed this reply, in review chat above, notifications in review chats always throw me off so I'm pulling it out here

> is the `last_hidden_state` would be equivalent to no pooled output?
> 
> where the last hidden state is:
> 
> ```python
> last_hidden_state, hidden_states = self.timm_model.forward_intermediates(pixel_values, **kwargs)
> # or
> last_hidden_state = self.timm_model.forward_features(pixel_values, **kwargs)
> ```
> 
> If so, I suppose we can add some flag to avoid applying `forward_head` if someone don't want to have pooler output. It can be the same`BaseModelOutputWithPooling` with `pooler_output=None`


Yes, last_hidden_state is always unpooled output, they are supposed to be equivalent btw `forward_intermediates` and `forward_features`, in both cases last_hidden_state in both cases should be possible to pass to forward_head() and produce the same result. 

You've already noticed, but forward_features exists in every single model and forward_intermediates only a few of them.

So for pooling is `BaseModelOutputWithPooling` .. is it normal to have `pooler_output=None`? Looking again at some convnets and looks like they always do include pooling output, I assumed that some might output BaseModelOutput but seems that's only intermediate stages. So maybe this is fine as is if that's typical in transformers.

Which brings to next point, I'm not intimately famaliar with the norms in Transformers, we should get another reviewer on here who is as my focus is on the timm bits.

qubvel commented:
> So for pooling is BaseModelOutputWithPooling .. is it normal to have pooler_output=None? Looking again at some convnets and looks like they always do include pooling output, I assumed that some might output BaseModelOutput but seems that's only intermediate stages. So maybe this is fine as is if that's typical in transformers.

Agreed, looks a bit weird. In case of confusion, we can create `TimmWrapperModelOutput` and correctly specify that `pooler_output` is optional based on the config/init option. I will add an option and let`s see the core maintainers' opinion.

qubvel committed:
Add do_pooling

qubvel committed:
Add test for do_pooling

qubvel committed:
Fix doc

qubvel committed:
Add tests for TimmWrapperModel

qubvel committed:
Add validation for `num_classes=0` in timm config + test for DINO checkpoint

qubvel committed:
Adjust atol for test

qubvel committed:
Fix docs

qubvel committed:
dev-ci

qubvel committed:
dev-ci

qubvel committed:
Add tests for image processor

qubvel committed:
Update docs

qubvel committed:
Update init to new format

qubvel committed:
Update docs in configuration

qubvel committed:
Fix some docs in image processor

qubvel committed:
Improve docs for modeling

qubvel committed:
fix for is_timm_checkpoint

qubvel committed:
Update code examples

qubvel committed:
Fix header

qubvel committed:
Fix typehint

qubvel reviewed (commented):
@LysandreJik please review whenever you have bandwidth!

Some tests are failing, waiting for the update of `timm` version in CI docker images in a separate PR.

qubvel committed:
Increase tolerance a bit

qubvel committed:
Merge branch 'main' into timm-wrapper

qubvel committed:
Fix Path

qubvel committed:
Fixing model parallel tests

qubvel committed:
Disable "parallel" tests

LysandreJik commented on src/transformers/utils/generic.py:
```diff
@@ -867,3 +869,48 @@ class LossKwargs(TypedDict, total=False):
     num_items_in_batch: Optional[int]
+
+
+def is_timm_hub_checkpoint(pretrained_model_name_or_path: str) -> bool:
```
Given the method's objective, I would have it accept only a `pretrained_model_name`

qubvel replied:
function removed with https://github.com/huggingface/transformers/commit/ff6efde8d0d77baa83e9e7a264aec84618751d1b (see comment below)

LysandreJik commented on src/transformers/utils/generic.py:
```diff
@@ -867,3 +869,48 @@ class LossKwargs(TypedDict, total=False):
+    if not isinstance(pretrained_model_name_or_path, str):
+        return False
+
+    if os.path.isfile(pretrained_model_name_or_path) or os.path.isdir(pretrained_model_name_or_path):
+        return False
+
```
and I'd therefore remove this

qubvel replied:
function removed with https://github.com/huggingface/transformers/commit/ff6efde8d0d77baa83e9e7a264aec84618751d1b

LysandreJik commented on src/transformers/utils/generic.py:
```diff
@@ -867,3 +869,48 @@ class LossKwargs(TypedDict, total=False):
+    if os.path.isfile(pretrained_model_name_or_path) or os.path.isdir(pretrained_model_name_or_path):
+        return False
+
+    return pretrained_model_name_or_path.startswith("hf-hub:timm/") or pretrained_model_name_or_path.startswith(
+        "timm/"
+    )
```
Good as long as we don't expect community checkpoints, but there are community checkpoints already, for example see the following: https://huggingface.co/prov-gigapath/prov-gigapath

I think we'll need a more robust check here

qubvel replied:
Indeed, this is a weak assumption. This function is only needed in the mage processors auto class to load it from the config. In `AutoImageProcessor.from_pretrained` the only information we have is the model name, so the only way to make a robust check is to load files from the Hub.

I removed this function entirely and instead made a fallback for loading the timm image processor dict in the auto class of the image processor. To avoid loading the config for every model, I did it in the following way:

1) Try to load the image processor config as usual - most of the models will be fine, and we won't have any overhead here.
2) In case of an exception, try loading config.json and check if it's a timm checkpoint.

See https://github.com/huggingface/transformers/pull/34564/commits/ff6efde8d0d77baa83e9e7a264aec84618751d1b for details. I documented it in the code, let me know if you have doubts about this approach.

Works with
```python
image_processor = AutoImageProcessor.from_pretrained("prov-gigapath/prov-gigapath")
```

LysandreJik replied:
This looks better to me!

LysandreJik commented on tests/models/timm_wrapper/test_modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,413 @@
+    def test_model_is_small(self):
+        pass
+
+    # Overriding as output_attentions is not supported by TimmWrapper
```
This should be removed no?

qubvel replied:
Thanks, removed in https://github.com/huggingface/transformers/pull/34564/commits/a476610cb5cc71e54975c47989d0ad1b680acfab

LysandreJik commented on tests/models/timm_wrapper/test_modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,413 @@
+
+@require_torch
+@require_vision
+class TimmWrapperModelIntegrationTest(unittest.TestCase):
```
Should require timm as well

qubvel replied:
Added in https://github.com/huggingface/transformers/pull/34564/commits/327095a8e120ff1dbbe06dbaf501fcdee8c6aa42

LysandreJik reviewed (commented):
This is starting to look nice!

qubvel committed:
Add comment for metadata

qubvel committed:
Refactor AutoImageProcessor for timm wrapper loading

qubvel committed:
Remove custom test_model_outputs_equivalence

qubvel committed:
Add require_timm decorator

qubvel committed:
Fix comment

qubvel commented on src/transformers/modeling_utils.py:
```diff
@@ -3971,7 +3907,9 @@ def from_pretrained(
 
-            if metadata.get("format") == "pt":
+            if metadata is None:
+                pass
```
Added https://github.com/huggingface/transformers/pull/34564/commits/7e0d2c6d5a32b42457ac9187a2f4d61972eee263

LysandreJik commented:
Thanks, this looks good! cc @molbap can you give the processor code a quick look just to double check?

molbap commented on src/transformers/models/auto/image_processing_auto.py:
```diff
@@ -413,7 +416,37 @@ def from_pretrained(cls, pretrained_model_name_or_path, *inputs, **kwargs):
+            image_processor_filename = IMAGE_PROCESSOR_NAME
+
+        # Load the image processor config
+        try:
+            # Main path for all transformers models and local TimmWrapper checkpoints
+            config_dict, _ = ImageProcessingMixin.get_image_processor_dict(
+                pretrained_model_name_or_path, image_processor_filename=image_processor_filename, **kwargs
+            )
+        except Exception as initial_exception:
+            # Fallback path for Hub TimmWrapper checkpoints. Timm models' image processing is saved in `config.json`
+            # instead of `preprocessor_config.json`. Because this is an Auto class and we don't have any information
+            # except the model name, the only way to check if a remote checkpoint is a timm model is to try to
+            # load `config.json` and if it fails with some error, we raise the initial exception.
+            try:
+                config_dict, _ = ImageProcessingMixin.get_image_processor_dict(
+                    pretrained_model_name_or_path, image_processor_filename=CONFIG_NAME, **kwargs
+                )
+            except Exception:
+                raise initial_exception
+
```
A remark here - that's true for any processor with a `preprocessor_config.json`, but would it make sense to sanitize the inputs a bit? On community checkpoints, there's extra keys that are unused, for instance `model_args` that contains duplicated information. 
(I'm asking because we're already using a try/catch pattern here so that allows some branching)

molbap replied:
basically I I load https://huggingface.co/prov-gigapath/prov-gigapath/blob/main/config.json, I'll end up with an `ImageProcessor` object that has keys I can't make much use of in transformers, such as `model_args`, so I wondered if it made sense to filter the contents of that `config.json` to an expected schema

qubvel replied:
Ok, it makes sense now. I made it on the image-processor init level, see https://github.com/huggingface/transformers/pull/34564/commits/fd7b646f3789413c75020fe2b7c6735c0dc8e9c4

Is it what you had in mind?

```python
>>> from transformers import AutoImageProcessor, AutoConfig

>>> image_processor = AutoImageProcessor.from_pretrained("prov-gigapath/prov-gigapath")
>>> print(image_processor)

TimmWrapperImageProcessor {
  "architecture": "vit_giant_patch14_dinov2",
  "data_config": {
    "crop_mode": "center",
    "crop_pct": 1.0,
    "input_size": [
      3,
      224,
      224
    ],
    "interpolation": "bicubic",
    "mean": [
      0.485,
      0.456,
      0.406
    ],
    "std": [
      0.229,
      0.224,
      0.225
    ]
  },
  "image_processor_type": "TimmWrapperImageProcessor"
}
```

molbap replied:
yes, exactly! that looks good

molbap commented on src/transformers/models/timm_wrapper/image_processing_timm_wrapper.py:
```diff
@@ -0,0 +1,116 @@
+
+        # If the input is a torch tensor, then no conversion is needed
+        # Otherwise, we need to pass in a list of PIL images
+        if isinstance(images, torch.Tensor):
+            images = self.val_transforms(images)
+            # Add batch dimension if a single image
+            images = images.unsqueeze(0) if images.ndim == 3 else images
```
here, if `val_transforms` is for instance
```python
Compose(
    Resize(size=224, interpolation=bicubic, max_size=None, antialias=True)
    CenterCrop(size=(224, 224))
    ToTensor()
    Normalize(mean=tensor([0.4850, 0.4560, 0.4060]), std=tensor([0.2290, 0.2240, 0.2250]))
)
```
Then the `ToTensor` op will fail, since `F.to_tensor` does expect a PIL image IIRC

qubvel replied:
Added in https://github.com/huggingface/transformers/pull/34564/commits/a10bc0de4f967f6361c8e1dc180a1e0aa8781174

molbap replied:
perfect!

rwightman replied:
@qubvel version checks tend to be brittle, can you do a hasattr on the MaybeToTensor class existing? I think that should line up with its use in the transforms?

qubvel replied:
@rwightman, thanks for the note. I added a check for the class name, it looks not that elegant but I hope it's more robust https://github.com/huggingface/transformers/pull/34564/commits/3d1a76e9fa1a6834ec5f99ecb0f2bb934fd37c19

molbap reviewed (commented):
Hey, took a quick look at the processor and ran it, found some stufff which I commented! Also looked at the whole PR, real nice work!

qubvel committed:
Make image processor work with older timm versions and tensor input

qubvel committed:
Save config instead of whole model in image processor tests

qubvel committed:
Add docstring for `image_processor_filename`

qubvel committed:
Sanitize kwargs for timm image processor

qubvel committed:
Fix doc style

qubvel committed:
Update check for tensor input

qubvel committed:
Merge branch 'main' into timm-wrapper

LysandreJik commented:
Ok awesome! At this point just a second quick look from @ArthurZucker and we're good

ArthurZucker commented on src/transformers/models/auto/configuration_auto.py:
very nice

qubvel replied:
thanks

ArthurZucker reviewed (approved):
Super nice ! This is kind of a perfect integration with Auto API, congrats!

qubvel committed:
Update normalize

qubvel committed:
Remove _load_timm_model function

qubvel commented on src/transformers/models/timm_wrapper/modeling_timm_wrapper.py:
```diff
@@ -0,0 +1,369 @@
+
+    def __init__(self, config: TimmWrapperConfig):
+        super().__init__(config)
+        self.timm_model = _load_timm_model(config, add_classification_head=False)
```
That sounds reasonable, removed in https://github.com/huggingface/transformers/pull/34564/commits/8b82e2b725f4bab53834afb5513b5b9b49031b6a

hongwhatamazon commented on src/transformers/image_processing_base.py:
```diff
@@ -285,6 +285,8 @@ def get_image_processor_dict(
             subfolder (`str`, *optional*, defaults to `""`):
                 In case the relevant files are located inside a subfolder of the model repo on huggingface.co, you can
                 specify the folder name here.
+            image_processor_filename (`str`, *optional*, defaults to `"config.json"`):
```
Isn't it defaults to "preprocessor_config.json"?

qubvel replied:
@hongwhatamazon, yes, you are right! Feel free to propose a PR to fix if you have bandwidt
